### ISSUE TITLE
Segfault when point cloud stats are nan

### ISSUE THREAD

**Issue by balazsdukai** [3DBAG/roofer#71]
Log:

```
[2024-11-14 10:59:45.458607708] INFO    Cropping pointcloud AHN3...
[2024-11-14 10:59:45.458623187] INFO    Mean point density = -nan
[2024-11-14 10:59:45.458627585] INFO    Standard deviation = -nan
[2024-11-14 10:59:45.458631532] INFO    Cropping pointcloud AHN4...
Segmentation fault (core dumped)
```

Test data and config: https://data.3dbag.nl/testdata/roofer/issue-71.zip

Command: `roofer --config roofer.toml`

Built from commit: 7cac1301b9db619b2edc0ad5d46a82834bf09cbe

**Comment by Ylannl** [comment]
I can't reproduce this on my Mac, will try on linux later.

**Comment by Ylannl** [comment]
Same on linux, it runs through, without showing the -nan.

**Comment by balazsdukai** [comment]
Maybe it is related to https://github.com/3DBAG/roofer/issues/64 ?

**Comment by balazsdukai** [comment]
Futher investigation shows that segfault only happens with the builtin logger. Spdlog does not segfault.
I'm not able reproduce the segfault on my laptop (gcc 13), but it is consistent on gilfoyle with gcc 11.
However, the original segfault in the docker image was also with gcc-13.
A debug run with the full input is not feasible, it never completes, not even on gilfoyle. I'm trying to get a subset that still segfaults.

**Comment by balazsdukai** [comment]
Finally I managed to run a small set in debug.
I'm getting an assertion violation at https://github.com/3DBAG/roofer/blob/7cac1301b9db619b2edc0ad5d46a82834bf09cbe/apps/roofer-app/roofer-app.cpp#L829 

The `tile.buildings` is empty, `tile.buildings_cnt` is 1, `tile.id` is 2.

Log:
```
[2024-11-18 07:36:38.749347650]	DEBUG	[cropper] Cropping tile 3
[2024-11-18 07:36:38.749386331]	DEBUG	[cropper] Tile extent: 103574.711 455546.797, 104574.711 456546.797
[2024-11-18 07:36:38.749459557]	DEBUG	[reconstructor] Submitted all buildings of tile 2 for reconstruction
[2024-11-18 07:36:38.749514829]	DEBUG	[reconstructor] after lock.unlock()
[2024-11-18 07:36:38.749556036]	DEBUG	[reconstructor] before lock cropped_tiles_mutex
[2024-11-18 07:36:38.749580781]	DEBUG	[reconstructor] before wait(lock)
[2024-11-18 07:36:38.749604916]	DEBUG	[reconstructor] crop_running.load() == true, !cropped_tiles.empty() == false
[2024-11-18 07:36:38.749693039]	DEBUG	[reconstructor] starting reconstruction for: output/issue-71/objects/NL.IMBAG.Pand.0499100002016291/reconstruct/NL.IMBAG.Pand.0499100002016291.city.jsonl
[2024-11-18 07:36:38.749838449]	DEBUG	[reconstructor] finished reconstruction for: output/issue-71/objects/NL.IMBAG.Pand.0499100002016291/reconstruct/NL.IMBAG.Pand.0499100002016291.city.jsonl
[2024-11-18 07:36:38.761362644]	INFO	Layer count: 1
[2024-11-18 07:36:38.764851676]	INFO	Layer 'reconstruction_input.t_7/236/192' total feature count: 316
[2024-11-18 07:36:38.764881792]	INFO	Getting layer extent...
[2024-11-18 07:36:38.764909303]	INFO	Layer extent: 102574.711 454546.797 103591.971 455570.745
[2024-11-18 07:36:38.764959125]	INFO	Layer 'reconstruction_input.t_7/236/192' total feature count: 316
[2024-11-18 07:36:38.764986105]	INFO	Layer geometry type: Polygon
[2024-11-18 07:36:38.765254442]	INFO	Setting spatial filter
[2024-11-18 07:36:38.766525877]	INFO	Simplifying and buffering footprints...
[2024-11-18 07:36:38.766589435]	INFO	Cropping pointcloud AHN3...
[2024-11-18 07:36:38.766659404]	INFO	Mean point density = -nan
[2024-11-18 07:36:38.766680353]	INFO	Standard deviation = -nan
[2024-11-18 07:36:38.766720307]	INFO	Analysing pointcloud AHN3...
[2024-11-18 07:36:38.766778615]	INFO	Selecting and writing pointclouds
[2024-11-18 07:36:38.767977656]	DEBUG	[cropper] Finished cropping tile 3, notifying reconstructor
[2024-11-18 07:36:38.768012921]	DEBUG	[cropper] Finished cropper
[2024-11-18 07:36:38.768089363]	DEBUG	[reconstructor] Submitted all buildings of tile 3 for reconstruction
[2024-11-18 07:36:38.768167177]	DEBUG	[reconstructor] after lock.unlock()
[2024-11-18 07:36:38.768221057]	DEBUG	[reconstructor] Waiting for all reconstructor threads to join...
[2024-11-18 07:36:38.768236756]	DEBUG	[sorter] after lock.unlock()
[2024-11-18 07:36:38.768426798]	DEBUG	[sorter] before lock reconstructed_buildings_mutex
[2024-11-18 07:36:38.768636406]	DEBUG	[serializer] after lock.unlock()
[2024-11-18 07:36:38.773867044]	DEBUG	[serializer] before lock sorted_tiles_mutex
[2024-11-18 07:36:38.911243543]	DEBUG	[reconstructor] finished reconstruction for: output/issue-71/objects/NL.IMBAG.Pand.0499100002014415/reconstruct/NL.IMBAG.Pand.0499100002014415.city.jsonl
[2024-11-18 07:36:38.944884653]	DEBUG	[reconstructor] finished reconstruction for: output/issue-71/objects/NL.IMBAG.Pand.0499100002016890/reconstruct/NL.IMBAG.Pand.0499100002016890.city.jsonl
[2024-11-18 07:36:40.702831605]	DEBUG	[reconstructor] finished reconstruction for: output/issue-71/objects/NL.IMBAG.Pand.1672100000015205/reconstruct/NL.IMBAG.Pand.1672100000015205.city.jsonl
[2024-11-18 07:36:41.393827082]	DEBUG	[reconstructor] finished reconstruction for: output/issue-71/objects/NL.IMBAG.Pand.1672100000015247/reconstruct/NL.IMBAG.Pand.1672100000015247.city.jsonl
[2024-11-18 07:36:42.597730433]	DEBUG	[reconstructor] finished reconstruction for: output/issue-71/objects/NL.IMBAG.Pand.1672100000015205/reconstruct/NL.IMBAG.Pand.1672100000015205.city.jsonl
[2024-11-18 07:36:45.217893540]	DEBUG	[reconstructor] finished reconstruction for: output/issue-71/objects/NL.IMBAG.Pand.0499100002013919/reconstruct/NL.IMBAG.Pand.049910000201391roofer: /home/bdukai/software/roofer-sync/apps/roofer-app/roofer-app.cpp:829: main(int, const char**)::<lambda()>: Assertion `tile.buildings.size() == tile.buildings_cnt' failed.
Signal: SIGABRT (Aborted)
```

**Comment by balazsdukai** [comment]
It seems that the segfault is caused by a nullptr in the BuildingTile.proj_helper, which can happen when the tile does not have any buildings.
0fdd6c2eb7b860122ca1d3b7ddc0e0a35a96f73c seems like a fix, however, the segfault is very elusive...
I added 3 tests with prefix `issue-71`, which are different configurations of the area that initially caused the segfault.
Once the fix is confirmed, it will be enough to keep the test `issue-71-v2`.
The branch https://github.com/3DBAG/roofer/tree/issue-71-baseline adds the tests on top of `develop`, so that the tests can be run against the base to see verify that the segfault happens.

**Comment by Ylannl** [comment]
I implemented a fix for this issue in #72. It is basically the same as what you proposed, but aborts a bit earlier in crop_tile(). An earlier fix in that PR fixes the 'next' problem with the `issue-71-v2` test, ie. that it hangs. See lines https://github.com/3DBAG/roofer/blob/3c827da164696bb15217f2f2dc5daff82c26087d/apps/roofer-app/roofer-app.cpp#L729-L730

With this all the issue-71 tests succeed for me on gilfoyle.

If you had any other changes in the issue-71 branch (the formatting of tiles to wkt?), can you put those in a new branch?
